### ISSUE TITLE
`onCompleted` and `onError` not working in `useMutation` execution function

### ISSUE THREAD

**Issue by dspacejs** [apollographql/apollo-client#8793]
Given the following code:

```ts
const [createMedicalDocument, { loading }] = useMutation(
  CREATE_MEDICAL_DOCUMENT_MUTATION,
  {
    onCompleted: (data) => console.log("runs on success"),
    onError: () => console.log("runs on error"),
  }
);

createMedicalDocument({
  onCompleted: () => console.log("doesn't run on success"),
  onError: () => console.log("doesn't run on error"),
});
```

The first set of `onCompleted` and `onError` runs, but not the second.

**Intended outcome:**
Both sets of `onCompleted` and `onError` should run.

**Actual outcome:**
Only the first set ran.

**How to reproduce the issue:**

https://github.com/dspacejs/react-apollo-error-template

In this example, neither set of `onSuccess` or `onError` runs. I couldn't reproduce my specific scenario exactly for some reason (where the first set runs, but not the second). 

**Versions**

```
  System:
    OS: Linux 5.11 Ubuntu 20.04.3 LTS (Focal Fossa)
  Binaries:
    Node: 14.17.1 - /usr/local/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 7.22.0 - /usr/local/bin/npm
  Browsers:
    Firefox: 92.0
  npmPackages:
    @apollo/client: ^3.4.8 => 3.4.10
```

**Comment by brainkim** [comment]
I’m sorry to hear you’re having trouble! So two things. First, in your reproduction you are using an `onSuccess`  option rather than an `onCompleted` option. Second, I don’t think the mutation function returned from the `useMutation()` hook takes additional parameters. So the `onCompleted` and `onError` callbacks are passed into the `useMutation()` hook only.

Maybe you were led astray by some incorrect documentation somewhere? One suggestion is that TypeScript type checking would have helped surface these issues a little more quickly!

**Comment by dspacejs** [comment]
@brainkim 

> First, in your reproduction you are using an `onSuccess` option rather than an `onCompleted` option.

Ah yes sorry that was a typo, in my real project (not the reproduction) I was using `onCompleted` (not `onSuccess`).

I've updated the reproduction accordingly. Now it works with the first set, but not the second (which is the same as my project). 

> Second, I don’t think the mutation function returned from the `useMutation()` hook takes additional parameters, or the `onCompleted`/`onError` options. So the `onCompleted` and `onError` callbacks are passed into the `useMutation()` hook only.

It does, I have TypeScript in my project and it has types for the `onSuccess` and `onError` for the function returned by `useMutation()`.

**Comment by brainkim** [comment]
Ah I guess the mutation execution function takes `onCompleted` and `onError` according to the types. My mistake! I’m not sure if that’s something we’ll support though, given that 1. as you noticed, the function returns a promise which provides essentially the same functionality and 2. there are still questions like should passing in `onCompleted` and `onError` overwrite any callbacks passed to the original hook?

**Comment by dspacejs** [comment]
@brainkim even if I remove the first set, the second set still never runs. So it's a bug AFAIK.

And regardless, if it enables you to add them twice, as a developer I expect it should run both sets (unless stated otherwise in the docs, and I've never seen it say this anywhere).

**Comment by brainkim** [comment]
@dspacejs I’m more leaning towards removing these as options from the mutate function in types, rather than adding support 😄

**Comment by dspacejs** [comment]
@brainkim I think you should add support for it. 

It works with `.then()` like I was saying, therefore I think there should also be the option for `onCompleted` and `onError`.

**Comment by dstoyanoff** [comment]
My 2 cents here are that the behavior should either be fixed or the TypeScript types should be corrected. The current behavior is leading to confusion and a lot of time spent in debugging.

Btw, docs say that the behavior, when passed in the mutate fn would be to override any initially set option - https://www.apollographql.com/docs/react/data/mutations/#option-precedence

**Comment by xvonabur** [comment]
Apollo client has `onCompleted` support  for `useQuery`, `useLazyQuery`, but not for `useMutation`? It's not a consistent API. It's confusing for me and others, so I think we definitely need `onCompleted` support for `useMutation`.

**Comment by brainkim** [comment]
Should be fixed in 3.5.4.

**Comment by latenightluke** [comment]
> 

I am on 3.5.6 and this still isn't working. :(
